kratos/geometries/nurbs_volume_geometry.h

```diff
@@ -0,0 +1,997 @@
+        }
+        else {
+            shape_function_container.ComputeBSplineShapeFunctionValues(mKnotsU, mKnotsV, mKnotsW,
+                rCoordinates[0], rCoordinates[1], rCoordinates[2]);
+        }
+
+        if (rResult.size() != shape_function_container.NumberOfNonzeroControlPoints())
+            rResult.resize(shape_function_container.NumberOfNonzeroControlPoints());
+
+        for (IndexType i = 0; i < shape_function_container.NumberOfNonzeroControlPoints(); ++i) {
+            rResult[i] = shape_function_container(i, 0);
+        }
+
+        return rResult;
+    }
+
+    Matrix& ShapeFunctionsLocalGradients(
+        Matrix& rResult,
+        const CoordinatesArrayType& rCoordinates) const override
+    {
+        NurbsVolumeShapeFunction shape_function_container(mPolynomialDegreeU, mPolynomialDegreeV, mPolynomialDegreeW, 1);
+
+        if (IsRational()) {
+            shape_function_container.ComputeNurbsShapeFunctionValues(mKnotsU, mKnotsV, mKnotsW, mWeights,
+                rCoordinates[0], rCoordinates[1], rCoordinates[2]);
+        }
+        else {
+            shape_function_container.ComputeBSplineShapeFunctionValues(mKnotsU, mKnotsV, mKnotsW,
+                rCoordinates[0], rCoordinates[1], rCoordinates[2]);
+        }
+
+        if (rResult.size1() != shape_function_container.NumberOfNonzeroControlPoints()
+            && rResult.size2() != 3)
+            rResult.resize(shape_function_container.NumberOfNonzeroControlPoints(), 3);
+
+        for (IndexType i = 0; i < shape_function_container.NumberOfNonzeroControlPoints(); ++i) {
+            rResult(i, 0) = shape_function_container(i, 1);
+            rResult(i, 1) = shape_function_container(i, 2);
+            rResult(i, 2) = shape_function_container(i, 3);
+        }
+
+        return rResult;
+    }
+
+    ///@}
+    ///@name Information
+    ///@{
+    std::string Info() const override
+    {
+        return std::to_string(TWorkingSpaceDimension) + " dimensional nurbs volume.";
+    }
+
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << TWorkingSpaceDimension << " dimensional nurbs volume.";
+    }
+
+    void PrintData(std::ostream& rOStream) const override
+    {
+    }
+    ///@}
+
+private:
+    ///@name Private Static Member Variables
+    ///@{
+
+    static const GeometryData msGeometryData;
+
+    static const GeometryDimension msGeometryDimension;
+
+    ///@}
+    ///@name Private Member Variables
+    ///@{
+
+    SizeType mPolynomialDegreeU;
+    SizeType mPolynomialDegreeV;
+    SizeType mPolynomialDegreeW;
+    Vector mKnotsU;
+    Vector mKnotsV;
+    Vector mKnotsW;
+    Vector mWeights;
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    /*
+    * @brief Checks if the knot vector is coinciding with the number of
+    *        control points and the polynomial degree. If the knot vectors
+    *        have a multiplicity of p+1 in the beginning, it is reduced to p.
+    */
```

Comment:
Why do we shrink p+1 to p?

Reply:
the thing is, some cad software give you p some gives p+1. However, by comparing the sizes one knows if it is p+1 or p. So we check that and store it in the way that we want. All algorithms are written in a way that it need p, ans it is a slice more memory efficient. Thus, we just check how the input is done and update it that we can use it correctly.